share/shwap/p2p/discovery/discovery.go

```diff
@@ -189,24 +189,16 @@ func (d *Discovery) Advertise(ctx context.Context) {
 
 			// we don't want retry indefinitely in busy loop
 			// internal discovery mechanism may need some time before attempts
-			errTimer := time.NewTimer(retryTimeout)
+			errTimer := time.After(retryTimeout)
 			select {
-			case <-errTimer.C:
-				errTimer.Stop()
-				if !timer.Stop() {
-					<-timer.C
-				}
+			case <-errTimer:
 				continue
 			case <-ctx.Done():
-				errTimer.Stop()
 				return
 			}
 		}
 
 		log.Infof("successfully advertised to topic %s", d.topic)
-		if !timer.Stop() {
-			<-timer.C
-		}
 		timer.Reset(d.params.AdvertiseInterval)
```

Comment:
Basing off 
```
// if the program has not received from t.C already and the timer is
// running, Stop is guaranteed to return true.
```

I think it should never get into the if case. I also believe they wouldn't break all the millions of loc written that were handling for this timer edge case this way.

Reply:
It will get to this case if timer didn't fire by the time Stop() was called.

Reply:
I think its the opposite actually:
>  It returns true if the call stops the timer, false if the timer has already
 expired or been stopped.
 

So in case `d.disc.Advertise(ctx, d.topic)` call takes longer than timer, meaning the timer ha fired by the time the call finishes.

Reply:
You right on correction for comment. But It was in PR description from the beginning:
```
In particular, If Timer.Stop is called after the Timer expires, we deadlock. This can happen if advertisement call or new block handling takes longer then the set deadline on the Timer, both are rare but possible case in production environment.
```

Reply:
Dude, that's the line I added, see the edit history of the PR